### PR TITLE
dm-relay: create binlog reader from relay, refactor relay

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve? 
part 2 of subtask 4 in [#2214 of DM](https://github.com/pingcap/dm/issues/2214)

i ever thought to split this pr, but it's too late after fixing all the test cases.

### What is changed and how it works?
- relay writer shouldn't be used concurrently and it's not a service, so remove unnecessary lock(except some status) and `stage` state
- remove unnecessary stage in binlog writer; reuse the same binlog writer in relay writer
- reuse relay writer in relay
- move `recover` logic from relay writer to relay
- remove unnecessary connection kill logic(already resolved in this [pr](https://github.com/go-mysql-org/go-mysql/pull/411), and current used version contains it)
- binlog reader created using relay; register listeners implicitly
- remove the transformer entity, make it a method in relay

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: dm/pkg/parser/common.go [hunk 1]
```diff
@@ -335,3 +335,27 @@ func SplitDDL(stmt ast.StmtNode, schema string) (sqls []string, err error) {
 func genTableName(schema string, table string) *filter.Table {
 	return &filter.Table{Schema: schema, Name: table}
 }
+
+// CheckIsDDL checks input SQL whether is a valid DDL statement.
+func CheckIsDDL(sql string, p *parser.Parser) bool {
+	sql = utils.TrimCtrlChars(sql)
+
+	if utils.IsBuildInSkipDDL(sql) {
```
Comment (lance6716) [inline dm/pkg/parser/common.go:343]: maybe address this behaviour "the default skipped DDL will return `false`" in function comment.

File: dm/relay/binlog_writer.go [hunk 2]
```diff
@@ -19,30 +19,27 @@ import (
...
 
-	"github.com/pingcap/ticdc/dm/pkg/binlog/common"
 	"github.com/pingcap/ticdc/dm/pkg/log"
 	"github.com/pingcap/ticdc/dm/pkg/terror"
 )
 
 // BinlogWriter is a binlog event writer which writes binlog events to a file.
 type BinlogWriter struct {
-	cfg *BinlogWriterConfig
+	mu sync.RWMutex
```
Comment (lance6716) [inline dm/relay/binlog_writer.go:32]: seems `FileWriter.out` is only used in one goroutine and driven by upstream binlog event, so we even need no lock.

**Review by lance6716 — COMMENTED** [review]
will review tests later

**Review by lichunzhu — COMMENTED** [review]
will submit some comments later

**Review by lichunzhu — APPROVED** [review]
lgtm

**Review by lance6716 — COMMENTED** [review]
rest lgtm, please address existing comments

File: dm/relay/binlog_writer.go (hunk 2 above)
Comment (D3Hunter) [reply dm/relay/binlog_writer.go:32]: "filename/offset" state'll be used in binlog reader later.

File: dm/pkg/parser/common.go (hunk 1 above)
Comment (D3Hunter) [reply dm/pkg/parser/common.go:343]: maybe another pr

File: dm/relay/relay.go [hunk 3]
```diff
@@ -123,17 +125,21 @@ type Relay struct {
...
 // NewRealRelay creates an instance of Relay.
 func NewRealRelay(cfg *Config) Process {
-	return &Relay{
+	r := &Relay{
 		cfg:       cfg,
 		meta:      NewLocalMeta(cfg.Flavor, cfg.RelayDir),
 		logger:    log.With(zap.String("component", "relay log")),
 		listeners: make(map[Listener]struct{}),
 	}
+	r.writer = NewFileWriter(r.logger)
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply dm/relay/relay.go:141]: it's using `r.logger`

File: dm/relay/binlog_writer.go [hunk 4]
```diff
@@ -19,30 +19,27 @@ import (
...
+	offset   atomic.Int64
+	file     *os.File
+	filename string
 
 	logger log.Logger
 }
 
 // BinlogWriterStatus represents the status of a BinlogWriter.
 type BinlogWriterStatus struct {
-	Stage    string `json:"stage"`
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply dm/relay/binlog_writer.go:45]: it over complicated for a file write wrapper to have a `stage`

whether it's opened or not can be inferred from `filename`

**Comment by D3Hunter** [comment]
@lance6716  PTAL

**Review by lance6716 — APPROVED** [review]

**Pull request merged**